src/libraries/System.Net.Primitives/ref/System.Net.Primitives.cs

```diff
@@ -323,6 +323,23 @@ public abstract partial class TransportContext
         protected TransportContext() { }
         public abstract System.Security.Authentication.ExtendedProtection.ChannelBinding? GetChannelBinding(System.Security.Authentication.ExtendedProtection.ChannelBindingKind kind);
     }
+    public enum NetworkError : int
+    {
+        Unknown = 0,
+        AddressInUse,
+        InvalidAddress,
+        HostNotFound,
+        ConnectionRefused,
+        ConnectionAborted,
+        ConnectionReset,
+    }
+    public class NetworkException : System.IO.IOException
+    {
+        protected NetworkException(NetworkError error, Exception? innerException = null, string? message = null) { }
+        protected NetworkException(System.Runtime.Serialization.SerializationInfo serializationInfo, System.Runtime.Serialization.StreamingContext streamingContext) { }
+        public NetworkError NetworkError { get { throw null; } }
+        public static NetworkException Create(NetworkError error, Exception? innerException = null, string? message = null) { throw null; }
```

Comment:
I wasn't in the API review. Why does this use protected ctors and a factory method? That's very uncommon with exceptions.

Reply:
The idea is that we could have derived exceptions like `AddressInUseException` later on, and that `Create` would be updated to throw those when we later add them.

Reply:
Why would we add such types? What additional information would the exception carry that would require a custom type?

Reply:
I think it's pretty unlikely we would ever add such types. The discussion in the meeting went around a bit and this may be an artifact of some intermediate stage of dicussion.

I'd be happy to just remove the factory method on the assumption that we'll never add those types. As you say, there doesn't really seem to be any reason to add them, since you already have all the info you need.

Reply:
Guided by @bartonjs guidance that exception filters should not be the primary method of distinguishing exceptions.

Reply:
> Guided by @bartonjs guidance that exception filters should not be the primary method of distinguishing exceptions that you would take different actions on.

Which we basically ended up ignoring in the end, though it took some discussion to get there.

Reply:
> Which we basically ended up ignoring in the end, though it took some discussion to get there.

Well, I think the expectation was that we'd come back with those concrete exception types (and that we would hold off on some we don't believe people will actually care to catch, like AddressInUse).

If we decide not to do those exception types then a factory method is meaningless.

Reply:
Ok, so do we actually want to do concrete exception types?

My vote is no, and remove the factory.

Reply:
> Ok, so do we actually want to do concrete exception types?

Yes.  If it's something that you think someone wants to handle in a catch it should (generally) be distinguished by type alone.

> My vote is no, and remove the factory.

Then it's probably worth bringing back to API review and having a different discussion.

Reply:
I'm happy to have more API review discussion on this, but I feel like the main issue in the discussion was that there didn't seem to be clear consensus from the API body itself about how to handle this. We started with exploring distinguished exception types, then gravitated away from that toward an enum, and then sort of seemed to close on that by default.

As you can probably tell, I don't really have a strong opinion here, I just want to do the right thing and move forward.

> If it's something that you think someone wants to handle in a catch

I think this is the rub, because the answer is basically, "maybe". We know ASP.NET has distinguished types for a few of these values (but not all), so that's a partial answer. We know that once we expose these in an enum or whatever, someone will invariably decide that they care about it for whatever reason, possibly good, possibly bad. But also, by far the common case is that people don't care about handling specific error conditions.

Reply:
> If it's something that you think someone wants to handle in a catch it should (generally) be distinguished by type alone

Why?

Reply:
The guidelines aren't 100% prescriptive here, but.

> CONSIDER creating and throwing custom exceptions if you have a program
error condition that can be programmatically handled in a
different way than any other existing exception. Otherwise, throw one
of the existing exceptions. ...

Which suggests "different handler, different custom exception"

> DO NOT create a new exception type if the exception would not be handled
differently than an existing framework exception. Throw the
existing framework exception in such case.

The contrapositive to this is: would be handled differently => new exception type

The only super-clear item (with regard to this specific point) is an annotation by @KrzysztofCwalina:

> It’s OK for exceptions to have a property
returning some kind of error code, but I would be very careful about this.
Each exception can carry two main pieces of information: the exception
message explaining to the developer what went wrong and how to fix it,
and the exception type that should be used by handlers to decide what programmatic
action to take. If you think you need to have a property on your
exception that would return additional error code, ask yourself who this
code is for. Is it for the developer or for the exception handler? If for the
developer, add additional information to the message. If for the handlers,
add a new exception type.

Reply:
I don't think the position follows from those CONSIDER/DO NOTs, and one person's opinion does not a policy make. SocketError, for example, has over 50 members; in no world would I want us adding over 50 exception types. On top of that, this guidance and that annotation were all written before exception filters were exposed and prominent in C#; it's now trivial to handle error code values like this, and in fact if there's any chance you might want to perform the same action for multiple such values, it's way simpler with filters than just going by exception type. Designing to the guidance of "avoid exception filters" is IMHO wrong.

Reply:
I'll give an argument for having distinct exception types:

We are no longer giving OS error codes with this API. We have distilled/merged similar error codes together, to the point where it really doesn't seem like someone would want to take action on multiple of them in a single catch block -- the actions would be so different that multiple catch blocks are warranted.

If that is true, then having multiple exception types makes more sense and would bring us a little closer to what I view as idiomatic C#, while also providing an error code leaves an escape hatch for anyone who would be in the rare situation of needing to combine multiple error codes.

Reply:
So you're saying the code that ASP.NET hashttps://source.dot.net/#Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/Internal/SocketConnection.cs,144 that currently uses an exception filter would stop doing so? I don't buy it, especially when factoring in that it's not only looking at multiple exception types, but multiple error codes conditioned by platform.

Reply:
@stephentoub we should consider using that logic here so that they don't have to. That should be a goal when reducing the `SocketError` into `NetworkError`.

Reply:
To me, it's about making sure that people can write their handlers correctly.

* If all errors are ignored, there's no catch, everything works.
* If all errors are retried, there's no filter, and the one exception type (which would be the base class for the specially singled-out ones) is sufficient.
* If a "common" (to whatever extent you want to call something common) pattern is to only care about (e.g.) ConnectionReset, then a distinct type has a better chance of producing code to match that desire than a filter.
  * If you leave off the filter, it still catches when you wanted, but also when you didn't.
  * If ConnectionReset ever gets more granular, you need to update your filter, but have no good way of knowing it, which makes an extensible-looking model be actually brittle.

In the discussion, it was suggested that ConnectionReset, ConnectionAborted, and possibly AddressInUse were the most likely things to be handled in a manner other than "any connection error".  Leading to the hybrid model:

* NetworkException : IOException
  * static NetworkException Create(NetworkError)
* enum NetworkError
* ConnectionResetException : NetworkException
* ConnectionAbortedException : NetworkException
* AddressInUseException : NetworkException

InvalidAddress is mostly a "something weird made it this far" (IIUC), so it seems (to me, and as suggested in the meeting) to not be something that would ever warrant a specialty type... because no one would really be handling it as a special case.  And while the other errors do have their particular uses (largely in painting UI in special ways for error handling) they might be handled rarely enough that they also aren't worth specialty types.

> SocketError, for example, has over 50 members; in no world would I want us adding over 50 exception type

While I'm not a sockets guru, I don't expect that there's really any recovery action for `NotSocket`; so it wouldn't warrant a specialty type (no special handler, no special type).  The fact that `NotSocket` happened is definitely useful for logging, so reporting the information is useful.  And maybe there's a super-power-scenario where someone can do something unique and clever, so also being programmatic isn't harmful.

Based on the suggested mappings for NetworkError from SocketError, many of the SocketError values are "similar, but different in nuance", and those would probably benefit/have-benefitted from a tiny bit of hierarchy, like `UnsupportedProtocolException : SocketException`, to represent ProtocolType/ProtocolOption/ProtocolNotSupported (I'm just guessing based on names).

> I don't think the position follows from those CONSIDER/DO NOTs

The conclusion that I draw from them is "everything that should be handled the same way in common usage of the API should be the same exception type", and "everything that should be handled in distinct ways in common usage of the API should be different exception types"; again, using robustness and the ability to evolve as the justification.  While Krzysztof's comment sort of suggests never having the programmatic properties, I don't quite get that from the guidelines.  I think that the hybrid model is a nice balance for "make easy things easy and hard things possible", or a variant "make common needs easy to fulfill, and uncommon needs possible to fulfill".

The impression I got from the meeting is ConnectionAborted and ConnectionReset are "commonly" handled, and are handled differently from each other, so different exception types are warranted.  The fact that the referenced ASP.Net code defined exceptions of exactly those names feels like reinforcement of that belief.

Reply:
> we should consider using that logic here so that they don't have to. That should be a goal when reducing the SocketError into NetworkError.

ASP.NET can choose to make such policy decisions as to how it reacts to various conditions. I don't think the same holds for the library. The comment there indicates it's only sometimes returning a different code than expected; we'd be applying a broken policy if we then always did such a translation for other consumers.

> idiomatic C#, 

And the next catch block that handles both an abort and a dispose in the same catch block? That's not "idiomatic C#"? It sure seems to me like a completely valid, prototypical use. If what you mean by idiomatic is "wouldn't have been written before C# 7", then sure, and we may as well stop investing in the language if we say anything beyond the initial releases isn't recommended.

I honestly don't understand the argument here for derived types. If two NetworkExceptions differ only in error code, and carry no different data otherwise, I see no good reason to introduce yet another exception type. It provides zero additional benefit from my perspective, only complication.